src/provider/index.ts

```diff
@@ -1,132 +1,114 @@
-import {AppId, Application, AppName} from '../client/directory';
-import {IntentType} from '../client/intents';
-import {RaiseIntentPayload} from '../client/internal';
-import {Context} from '../client/main';
-
-import {FDC3} from './FDC3';
-import {AppMetadata} from './MetadataStore';
-
-console.log('the provider has landed.');
-
-// Create and initialise desktop agent
-const service: FDC3 = new FDC3();
-service.register();
-
-/**
- * When the user is shown an app selector popup, they have the option of telling
- * the service how to handle similar intents in the future.
- *
- * This enum defines the options available to users.
- */
-export const enum DefaultAction {
-    /**
-     * Service should always show the app selection UI, to allow the user to
-     * choose which application to use.
-     */
-    ALWAYS_ASK = 'ALWAYS_ASK',
-
-    /**
-     * The service should always use the current selection when the intent is
-     * coming from the app that fired the current intent.
-     */
-    ALWAYS_FOR_APP = 'ALWAYS_FOR_APP',
-
-    /**
-     * The service should always use the current selection, whenever an intent of
-     * this type is fired.
-     */
-    ALWAYS_FOR_INTENT = 'ALWAYS_FOR_INTENT'
+import 'reflect-metadata';
+
+import {RaiseIntentPayload, APITopic, OpenPayload, FindIntentPayload, FindIntentsByContextPayload, BroadcastPayload} from '../client/internal';
+import {AppIntent, IntentResolution, Application, Intent} from '../client/main';
+
+import {injectable, inject} from 'inversify';
+import {Inject} from './common/Injectables';
+import {AppDirectory} from './model/AppDirectory';
+import {Model, FindFilter} from './model/Model';
+import {ContextHandler} from './controller/ContextHandler';
+import {IntentHandler} from './controller/IntentHandler';
+import {APIHandler} from './APIHandler';
+import {Injector} from './common/Injector';
+
+interface API {
+    [APITopic.OPEN]: [OpenPayload, void];
+    [APITopic.FIND_INTENT]: [FindIntentPayload, AppIntent];
+    [APITopic.FIND_INTENTS_BY_CONTEXT]: [FindIntentsByContextPayload, AppIntent[]];
+    [APITopic.BROADCAST]: [BroadcastPayload, void];
+    [APITopic.RAISE_INTENT]: [RaiseIntentPayload, IntentResolution];
 }
 
-// Message definitions
-export interface OpenArgs {
-    name: AppName;
-    context?: Context;
-}
-export interface ResolveArgs {
-    intent: IntentType;
-    context?: Context;
-}
-export interface SelectorResultArgs {
-    handle: number;
-    success: boolean;
-
-    /**
-     * The application that was selected by the user.
-     *
-     * Only specified when success is true.
-     */
-    app?: Application;
-
-    /**
-     * The reason that an app wasn't selected.
-     *
-     * Only specified when success is false.
-     */
-    reason?: string;
-
-    /**
-     * Determines the future behaviour of this intent
-     */
-    defaultAction: DefaultAction;
+@injectable()
+export class Main {
```

Comment:
Can we find a better name than "Main"? Maybe something like "FDC3Service", but open to any suggestions?

Reply:
Sure, can discuss. Bugs me a bit that the entry point here is `index.ts`, but in the client it's `main.ts`. Should pick a consistent pattern.

Reply:
Agreed